### PR TITLE
Respect options(warn = -1), fixes #958

### LINKED ISSUE #958: Testthat reporting warnings that do not appear in code
## Minimal reprex 

from @gaborcsardi 

```r
test_that("x", {
  op <- options(warn = -1)  
  warning("not see this")
  options(op)
  expect_true(TRUE)
})
```

## Real world case

If you run this in R, there is no warning:

```r
out <- available.packages(repos = 'https://cloud.r-project.org', type = 'mac.binary.el-capitan')
```

However if you run this code anywhere in testthat:

Screen Shot 2019-11-04 at 8 44 45 PM

This warning should never appear because [R sets `options(warn = -1L)` internally](https://github.com/wch/r-source/blob/c85acc47b4c3857b5a72f1871053bb5c90f2d08a/src/library/utils/R/packages.R#L87-L93).

### PR DESCRIPTION
**Pull request by jeroen**
I don't know how to test this. Is there something like `expect_no_warning()`?

### CODE REVIEW

**Comment by hadley** [comment]
`expect_warning(foofy(), NA)`

(can you please also merge/rebase to resolve some of the spurious check failures?)

File: NEWS.md [hunk 1]
```diff
@@ -34,6 +34,8 @@ This release mostly focusses on an overhaul of how testthat works with condition
   should be lazily registered, e.g. with `vctrs::s3_register()`. This
   is useful for introducing an experimental error class without
   encouraging users to depend on the class in their tests.
+  
+* Testthat now respects options(warn = -1) to ignore all warnings.
```
Comment (hadley) [inline NEWS.md]: Can you please acknowledge yourself and the issue as in https://style.tidyverse.org/news.html#acknowledgement

**Comment by jeroen** [comment]
I'm so confused. This PR fixed my real world problem, but the unit tests still fails :/  Would you mind having a look to find a solution that satisfies the test?

**Comment by jeroen** [comment]
The confusing thing is that the bug actually does not appear when in `expect_warning()` but in the reported when warnings were raised but not caught.

**Comment by hadley** [comment]
The way this works is nightmarishly complex, so I think this will have to wait until this release, unfortunately.

**Comment by jeroen** [comment]
OK well the fix in this PR at least fixes the reporter problem about uncaught warnings. So I think it's definitely an improvement, and maybe it can be merged?

I'll remove the test, because this actually tests something else than the warning reporting issue, namely if `expect_warning` correctly deals with options.

File: R/test-that.R [hunk 2]
```diff
@@ -115,9 +115,10 @@ test_code <- function(test, code, env = test_env(), skip_on_empty = TRUE) {
     invokeRestart("continue_test")
   }
   handle_warning <- function(e) {
+    # When options(warn) < 0, warnings are expected to be ignored.
     # When options(warn) >= 2, a warning will be converted to an error.
     # So, do not handle it here so that it will be handled by handle_error.
-    if (getOption("warn") >= 2) {
+    if (getOption("warn") < 0 || getOption("warn") >= 2) {
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/test-that.R]: Actually, I'm going to take a look now since this PR is literally the only thing standing in the way of a testthat release 😄

**Comment by hadley** [comment]
Thanks!

**Comment by jeroen** [comment]
Awesome thanks!

**Pull request merged**